### PR TITLE
138676271 post account memberships v5

### PR DESCRIPTION
**Pull request by wlffann**
In implementing the `POST /accounts/:account_id/memberships` endpoint, this PR creates a Proxy class to handle the chained methods on `AccountMembership`. 

Merge conflicts need to be resolved.

[#138676271]

### CODE REVIEW

File: lib/pivotal-tracker/account_membership.rb [hunk 1]
```diff
@@ -0,0 +1,27 @@
+class PivotalTracker::AccountMembership
+  attr_accessor :kind, :id, :person, :account_id,
```
Comment (tmikeschu) [inline lib/pivotal-tracker/account_membership.rb]: @wlffann I think these should be alphabetized, to follow the existing templates.

File: lib/pivotal-tracker/account.rb [hunk 2]
```diff
@@ -0,0 +1,23 @@
+class PivotalTracker::Account
+  attr_accessor :created_at, :days_left, :id, :kind,
```
Comment (tmikeschu) [inline lib/pivotal-tracker/account.rb]: @wlffann the order for these will get updated if we merge #5 first.

File: lib/pivotal-tracker/account_membership.rb [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  end
+
+  def self.all(account_id)
+    response = PivotalTracker.get("/accounts/#{account_id}/memberships")
+    if response.code == 200
+      response.parsed_response.map do |membership|
+        PivotalTracker::AccountMembership.new(membership)
+      end
+    else
+      puts response.inspect
```
Comment (tmikeschu) [inline lib/pivotal-tracker/account_membership.rb]: @wlffann if we merge #5 first, this error should grab the constant that will be in the PivotalTracker model.

File: lib/pivotal-tracker/account_membership.rb.orig [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+        PivotalTracker::AccountMembership.new(membership)
+      end
+    else
+      puts response.inspect
+    end
+  end
+
+  def self.create(membership_attributes, owner_id)
+    response = PivotalTracker.post("/accounts/#{owner_id}/memberships", query: membership_attributes)
+<<<<<<< HEAD
```
Comment (tmikeschu) [inline lib/pivotal-tracker/account_membership.rb.orig]: @wlffann merge conflicts! I see this has a `.orig` extension. Maybe just delete the files with that extension?

File: lib/pivotal-tracker/proxy.rb [hunk 5]
```diff
@@ -0,0 +1,29 @@
+class BasicObject #:nodoc:
```
Comment (tmikeschu) [inline lib/pivotal-tracker/proxy.rb]: @wlffann Killer trailblazing on this!

File: spec/lib/pivotal-tracker/account_spec.rb [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+  end
+
+  after do
+    VCR.eject_cassette
+  end
+
+  describe "#find" do
+    let(:account) { subject.find(account_id) }
+    context "with a valid account id" do
+      let(:account_id) { 962655 }
```
Comment (tmikeschu) [inline spec/lib/pivotal-tracker/account_spec.rb]: @wlffann I had to replace the hard-coded id's with `ENV['PIVOTAL_ACCOUNT_ID']`

File: lib/pivotal-tracker/proxy.rb.orig [hunk 7]
```diff
@@ -0,0 +1,41 @@
+class BasicObject #:nodoc:
```
Comment (tmikeschu) [inline lib/pivotal-tracker/proxy.rb.orig]: Same as comment above about deleting these `.orig` files.

**Review by tmikeschu — COMMENTED** [review]
Stellar work! 👌

File: lib/pivotal-tracker/proxy.rb [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+  def create(args = {})
+    @target.create(args, owner.id)
+  end
+
+  private
+    def method_missing(method, *args, &block)
+      @target.send(method, *args, &block)
+    end
+end
```
Comment (kheppenstall) [inline lib/pivotal-tracker/proxy.rb]: @wlffann With the `attr_reader` do we need all the `@` symbols in this file?

File: lib/pivotal-tracker/proxy.rb (hunk 5 above)
Comment (lucyconklin) [reply lib/pivotal-tracker/proxy.rb]: @wlffann @tmikeschu Yes! Thanks for tackling this. It seems like a pretty slick way to handle objects.

File: lib/pivotal-tracker/account.rb (hunk 2 above)
Comment (jasonnoble) [reply lib/pivotal-tracker/account.rb]: 👍

File: lib/pivotal-tracker/account_membership.rb (hunk 1 above)
Comment (jasonnoble) [reply lib/pivotal-tracker/account_membership.rb]: Agreed.

File: lib/pivotal-tracker/account_membership.rb.orig [hunk 9]
```diff
@@ -0,0 +1,31 @@
+class PivotalTracker::AccountMembership
```
Comment (jasonnoble) [inline lib/pivotal-tracker/account_membership.rb.orig]: `.orig` files are a side affect of `git mergetool` and should not be committed.  This is usually caused by running `git add .` instead of specifically staging the changes you want staged.

Please `git rm` this file, commit, then squash it into the existing commits.

File: lib/pivotal-tracker/account_membership.rb.orig (hunk 4 above)
Comment (jasonnoble) [reply lib/pivotal-tracker/account_membership.rb.orig]: Yup.  Need to remove them.

File: lib/pivotal-tracker/proxy.rb (hunk 5 above)
Comment (jasonnoble) [reply lib/pivotal-tracker/proxy.rb]: Do we need the `BasicObject` class?  See built in one:  https://ruby-doc.org/core-2.2.0/BasicObject.html

Comment (jasonnoble) [reply lib/pivotal-tracker/proxy.rb]: https://ruby-doc.org/stdlib-2.2.1/libdoc/delegate/rdoc/SimpleDelegator.html

File: lib/pivotal-tracker/account_membership.rb.orig (hunk 4 above)
Comment (wlffann) [reply lib/pivotal-tracker/account_membership.rb.orig]: Right! Thank you!

File: lib/pivotal-tracker/proxy.rb [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+class PivotalTracker::Proxy < BasicObject
+  attr_reader :owner, :target
+
+  def initialize(owner, target)
+    @owner = owner
+    @target = target
+  end
+
+  def all
+    @target.all(@owner.id)
```
Comment (jasonnoble) [inline lib/pivotal-tracker/proxy.rb]: `target`

File: lib/pivotal-tracker/proxy.rb [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+  end
+
+  def find(param)
+    @target.find(param, @owner.id)
```
Comment (jasonnoble) [inline lib/pivotal-tracker/proxy.rb]: `target`

File: lib/pivotal-tracker/proxy.rb [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+  end
+
+  def create(args = {})
+    @target.create(args, owner.id)
```
Comment (jasonnoble) [inline lib/pivotal-tracker/proxy.rb]: `target`

File: lib/pivotal-tracker/proxy.rb [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+  end
+
+  private
+    def method_missing(method, *args, &block)
+      @target.send(method, *args, &block)
```
Comment (jasonnoble) [inline lib/pivotal-tracker/proxy.rb]: `target`

File: lib/pivotal-tracker/proxy.rb (hunk 8 above)
Comment (jasonnoble) [reply lib/pivotal-tracker/proxy.rb]: 👍

File: lib/pivotal_tracker.rb [hunk 14]
```diff
@@ -7,6 +7,9 @@ class PivotalTracker
 require 'pivotal-tracker/client'
 require 'pivotal-tracker/story'
 require 'pivotal-tracker/project'
+require 'pivotal-tracker/account'
```
Comment (jasonnoble) [inline lib/pivotal_tracker.rb]: Please switch to `autoload`

File: spec/lib/pivotal-tracker/account_spec.rb (hunk 6 above)
Comment (jasonnoble) [reply spec/lib/pivotal-tracker/account_spec.rb]: Rebasing might fix?

File: spec/lib/pivotal-tracker/account_spec.rb [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+    end
+
+    it "all returned objects are memberships" do
+      account.memberships.all.each do |membership|
+        expect(membership).to be_a(PivotalTracker::AccountMembership)
+      end
+    end
+
+    it "a new membership can be created" do
+      new_membership = account.memberships.create(:email => "lucy.conklin@gmail.com", :name => "Lucy Conklin", :initials => "LC")
```
Comment (jasonnoble) [inline spec/lib/pivotal-tracker/account_spec.rb]: Use `Faker` gem to generate a name/email/initials?  Then verify the new_membership fields?

File: spec/lib/pivotal-tracker/proxy.rb [hunk 16]
```diff
@@ -0,0 +1,11 @@
+class Proxy
```
Comment (jasonnoble) [inline spec/lib/pivotal-tracker/proxy.rb]: https://ruby-doc.org/stdlib-2.2.1/libdoc/delegate/rdoc/SimpleDelegator.html ?

File: README.md.orig [hunk 17]
```diff
@@ -0,0 +1,73 @@
+## pivotal-tracker.rb - the original ruby gem
```
Comment (jasonnoble) [inline README.md.orig]: Please remove me...

File: lib/pivotal-tracker/account.rb.orig [hunk 18]
```diff
@@ -0,0 +1,38 @@
+class PivotalTracker::Account
```
Comment (jasonnoble) [inline lib/pivotal-tracker/account.rb.orig]: Remove me...

File: lib/pivotal-tracker/account_membership.rb (hunk 1 above)
Comment (wlffann) [reply lib/pivotal-tracker/account_membership.rb]: Got it! Thanks

File: lib/pivotal-tracker/account_membership.rb [hunk 19]
```diff
@@ -1,6 +1,29 @@
-class PivotalTracker::AccountMembership < PivotalTracker::Resource
+class PivotalTracker::AccountMembership
```
Comment (jasonnoble) [inline lib/pivotal-tracker/account_membership.rb]: Please undo this change.

File: lib/pivotal-tracker/account_membership.rb [hunk 20]
```diff
@@ -1,6 +1,29 @@
...
+      response.parsed_response.map do |membership|		
+        PivotalTracker::AccountMembership.new(membership)		
+      end		
+    else		
+      raise(PivotalTracker::PermissionDenied, "Only Admins and Owners can see account memberships")		
+    end		
+  end
+
+  def self.create(membership_attributes, account_id)
+    response = PivotalTracker.post("/accounts/#{account_id}/memberships", query: membership_attributes)
```
Comment (jasonnoble) [inline lib/pivotal-tracker/account_membership.rb:19]: Should we check for a 200 here?  (Like line 16)

File: lib/pivotal-tracker/proxy.rb [hunk 21]
```diff
@@ -0,0 +1,17 @@
+class PivotalTracker::Proxy
```
Comment (jasonnoble) [inline lib/pivotal-tracker/proxy.rb:1]: Inherit from resource?

Comment (jasonnoble) [inline lib/pivotal-tracker/proxy.rb:1]: Is this a general purpose proxy?  Or is it specific to Account Memberships?

File: lib/pivotal_tracker.rb [hunk 22]
```diff
@@ -1,9 +1,9 @@
-require 'httparty'
-
-class PivotalTracker
-
+require 'httparty'
```
Comment (jasonnoble) [inline lib/pivotal_tracker.rb]: Unrelated change?

File: lib/pivotal_tracker.rb [hunk 23]
```diff
@@ -13,11 +13,13 @@ class PivotalTracker
 PivotalTracker.autoload :Resource,          'pivotal-tracker/resource'
 PivotalTracker.autoload :JiraIntegration, 'pivotal-tracker/jira_integration'
 PivotalTracker.autoload :Me,                'pivotal-tracker/me'
+ PivotalTracker.autoload :AccountMembership, 'pivotal-tracker/account_membership'		
+ PivotalTracker.autoload :Proxy,             'pivotal-tracker/proxy'
 
-class PivotalTracker
-  include HTTParty
+class PivotalTracker
```
Comment (jasonnoble) [inline lib/pivotal_tracker.rb]: Remove trailing white space...

File: spec/lib/pivotal-tracker/account_spec.rb [hunk 24]
```diff
@@ -1,52 +1,59 @@
...
 
-    it "all returned objects are memberships" do
-      account.memberships.each do |membership|
-        expect(membership).to be_a(PivotalTracker::AccountMembership)
-      end
+      expect(new_membership).to be_a(PivotalTracker::AccountMembership)
     end
-  end
-end
+  end
```
Comment (jasonnoble) [inline spec/lib/pivotal-tracker/account_spec.rb]: Trailing white space

File: lib/pivotal-tracker/proxy.rb (hunk 21 above)
Comment (wlffann) [reply lib/pivotal-tracker/proxy.rb:1]: It is a general purpose Proxy. Any nested accounts can use to to called chained methods

Comment (wlffann) [reply lib/pivotal-tracker/proxy.rb:1]: Its initialize process does not take a hash of attributes, so I don't believe that it should inherit from resource

File: README.md [hunk 25]
```diff
@@ -66,4 +66,4 @@ The API is based on the following gist: http://gist.github.com/283120
 
 * Justin Smestad (https://github.com/jsmestad)
 * Jon Mischo (https://github.com/supertaz)
-* Gabor Ratky (https://github.com/rgabo)
\ No newline at end of file
+* Gabor Ratky (https://github.com/rgabo)
```
Comment (jasonnoble) [inline README.md]: Unrelated change...

File: lib/pivotal-tracker/account.rb [hunk 26]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (jasonnoble) [inline lib/pivotal-tracker/account.rb]: Remove me

File: lib/pivotal_tracker.rb [hunk 27]
```diff
@@ -6,6 +6,7 @@ class PivotalTracker
   base_uri 'https://www.pivotaltracker.com/services/v5'
 
   PermissionDenied = Class.new(Error)
+
```
Comment (jasonnoble) [inline lib/pivotal_tracker.rb]: Unrelated...

**Review by jasonnoble — CHANGES_REQUESTED** [review]

File: lib/pivotal-tracker/account_membership.rb [hunk 28]
```diff
@@ -3,4 +3,24 @@ class PivotalTracker::AccountMembership < PivotalTracker::Resource
...
+    if response.code == 200
+      new_account = PivotalTracker::AccountMembership.new(response.parsed_response)
+    else
+      raise(PivotalTracker::PermissionDenied, "Only Admins and Owners can add account memeberships")
```
Comment (jasonnoble) [inline lib/pivotal-tracker/account_membership.rb]: `memeberships` vs `memberships`
